src/mlib/str.h

```diff
@@ -0,0 +1,767 @@
+ * array (not including the null terminator). These two members should not be
+ * modified.
+ *
+ * The `view` member is a union member that will view the `mstr` as an
+ * @ref mstr_view.
+ *
+ * If you create an @ref mstr, it MUST eventually be passed to @ref mstr_free()
+ *
+ * The pointed-to code units of an mstr are immutable. To initialize the
+ * contents of an mstr, @ref mstr_new returns an @ref mstr_mut, which can then
+ * be "sealed" by converting it to an @ref mstr through the @ref mstr_mut::mstr
+ * union member.
+ *
+ * By convention, passing/returning an `mstr` to/from a function should
+ * relinquish ownership of that `mstr` to the callee/caller, respectively.
+ * Passing or returning an `mstr_view` is non-owning.
+ */
+typedef struct mstr {
+   union {
+      struct {
+         /**
+          * @brief Pointer to the beginning of the code unit array.
+          *
+          * @note DO NOT MODIFY
+          */
+         const char *data;
+         /**
+          * @brief Length of the pointed-to code unit array
+          *
+          * @note DO NOT MODIFY
+          */
+         size_t len;
+      };
+      /**
+       * @brief A non-owning `mstr_view` of the string
+       */
+      mstr_view view;
+   };
+} mstr;
+
+/**
+ * @brief An interface for initializing the contents of an mstr.
+ *
+ * Returned by @ref mstr_new(). Once initialization is complete, the result can
+ * be used as an @ref mstr by accessing the @ref mstr_mut::mstr member.
+ */
+typedef struct mstr_mut {
+   union {
+      struct {
+         /**
+          * @brief Pointer to the beginning of the mutable code unit array.
+          *
+          * @note DO NOT MODIFY THE POINTER VALUE. Only modify the pointed-to
+          * characters.
+          */
+         char *data;
+         /**
+          * @brief Length of the pointed-to code unit array.
+          *
+          * @note DO NOT MODIFY
+          */
+         size_t len;
+      };
+      /// Convert the mutable string to an immutable string
+      struct mstr mstr;
+      /// Convert the mutable string to an immutable string view
+      mstr_view view;
+   };
+} mstr_mut;
+
+/**
+ * @brief A null @ref mstr
+ */
+#define MSTR_NULL ((mstr){{{.data = NULL, .len = 0}}})
+/**
+ * @brief A null @ref mstr_view
+ */
+#define MSTRV_NULL ((mstr_view){.data = NULL, .len = 0})
+
+/**
+ * @brief Create an @ref mstr_view that views the given string literal
+ */
+#define mstrv_lit(String) (mstrv_view_data (String "", (sizeof String) - 1))
+
+/**
+ * @brief Create a new mutable code-unit array of the given length,
+ * zero-initialized. The caller can then modify the code units in the array via
+ * the @ref mstr_mut::data member. Once finished modifying, can be converted to
+ * an immutable mstr by copying the @ref mtsr_mut::mstr union member.
+ *
+ * @param len The length of the new string.
+ * @return mstr_mut A new mstr_mut
+ *
+ * @note The @ref mstr_mut::mstr member MUST eventually be given to
+ * @ref mstr_free().
+ */
```

Comment:
Who is responsible for handling allocation failures?

Reply:
Failure to allocate is not well handled in libmongocrypt (MONGOCRYPT-208).